pipelines/facebook_ads/__init__.py

```diff
@@ -0,0 +1,221 @@
+    account = get_ads_account(account_id, access_token, request_timeout)
+
+    def _get_data_chunked(method: TFbMethod, fields: Sequence[str], states: Sequence[str]) -> Iterator[TDataItems]:
+        # add pagination and chunk into lists
+        params: DictStrAny = {"limit": chunk_size}
+        if states:
+            params.update({'effective_status': states})
+        it: map[DictStrAny] = map(lambda c: c.export_all_data(), method(fields=fields, params=params))  # type: ignore
+        while True:
+            chunk = list(itertools.islice(it, chunk_size))
+            if not chunk:
+                break
+            yield chunk
+
+    @dlt.resource(primary_key="id", write_disposition="replace")
+    def campaigns(fields: Sequence[str] = DEFAULT_CAMPAIGN_FIELDS, states: Sequence[str] = None) -> Iterator[TDataItems]:
+        yield _get_data_chunked(account.get_campaigns, fields, states)
+
+    @dlt.resource(primary_key="id", write_disposition="replace")
+    def ads(fields: Sequence[str] = DEFAULT_AD_FIELDS, states: Sequence[str] = None) -> Iterator[TDataItems]:
+        yield _get_data_chunked(account.get_ads, fields, states)
+
+    @dlt.resource(primary_key="id", write_disposition="replace")
+    def ad_sets(fields: Sequence[str] = DEFAULT_ADSET_FIELDS, states: Sequence[str] = None) -> Iterator[TDataItems]:
+        yield _get_data_chunked(account.get_ad_sets, fields, states)
+
+    @dlt.transformer(primary_key="id", write_disposition="replace", selected=True)
+    def leads(items: TDataItems, fields: Sequence[str] = DEFAULT_LEAD_FIELDS, states: Sequence[str] = None) -> Iterator[TDataItems]:
+        for item in items:
+            ad = Ad(item["id"])
+            yield _get_data_chunked(ad.get_leads, fields, states)
+
+    @dlt.resource(primary_key="id", write_disposition="replace")
+    def ad_creatives(fields: Sequence[str] = DEFAULT_ADCREATIVE_FIELDS, states: Sequence[str] = None) -> Iterator[TDataItems]:
+        yield _get_data_chunked(account.get_ad_creatives, fields, states)
+
+    return campaigns, ads, ad_sets, ad_creatives, ads | leads  # type: ignore
+
+
+def enrich_ad_objects(fb_obj_type: AbstractObject, fields: Sequence[str]) -> ItemTransformFunctionWithMeta[TDataItems]:
+    """Returns a transformation that will enrich any of the resources returned by `` with additional fields
+
+    In example below we add "thumbnail_url" to all objects loaded by `ad_creatives` resource:
+    >>> fb_ads = facebook_ads_source()
+    >>> fb_ads.ad_creatives.add_step(enrich_ad_objects(AdCreative, ["thumbnail_url"]))
+
+    Internally, the method uses batch API to get data efficiently. Refer to demo script for full examples
+
+    Args:
+        fb_obj_type (AbstractObject): A Facebook Business object type (Ad, Campaign, AdSet, AdCreative, Lead). Import those types from this module
+        fields (Sequence[str]): A list/tuple of fields to add to each object.
+
+    Returns:
+        ItemTransformFunctionWithMeta[TDataItems]: _description_
```

Comment:
`_description_`: It looks like docs generation placeholder has been left unrendered.